[TECH] Utilisation d'un `CalibratedChallenge` dans la récupération d'un challenge (PIX-17946).

alexandrecoin opened the pull request:
## 🍂 Problème

En plus de ne pas refléter l'aspect métier de choix d'un Challenge via différentes propriétés importantes pour cette opération, le choix d'un challenge en certification est une opération se révélant assez couteuse en mémoire, qu'il est possible d'améliorer.

En effet, lors de toutes les opérations réalisées dans le usecase `get-next-challenge.js`, un tableau de `Challenge`s (contenant de base l'intégralité des challenges compatibles avec l'algo V3 - ayant un discriminant et une difficulté non nuls)
est récupéré, filtré, dupliqué, etc. créant une charge mémoire non utile.

## 🌰 Proposition

Pour pallier à ces deux problèmes, nous avons créé deux modèles `CalibratedChallenge` et `CalibratedChallengeSkill`, version allégées des modèles `Challenge` et `Skill` partagés dans tout Pix.

Ces modèles contiennent uniquement les données nécessaires au usecase `get-next-challenge.js` incluant :

- Des filtres par rapport à divers paramètres (ex: accessibilité)
- L'algorithme de sélection de challenges (⚠️  mention spéciale aux règles applicables dans `FlashAlgorithmRuleEngine.js`)

Une fois le challenge final sélectionné dans sa version `CalibratedChallenge`, celui-ci est "hydraté" avec les paramètres manquants par rapport à ceux du modèle `Challenge` et renvoyé à l'utilisateur.

## 🍁 Remarques

Cette PR est une première itération de ce à quoi pourrait ressembler cette selection de challenge "allégée" et certains points sont possiblement à revoir.

🧪 Les tests unitaires du usecase `get-next-challenge_test.js` sont une horreur pour la maintenance, seule l'implémentation y est testée.
Ils ont été mis à jour dans cette PR pour ne pas la surcharger mais mériteraient une PR dédiée pour passer d'une série de tests unitaires à des tests d'intégration.

## 🪵 Pour tester

- Passer une session de certification Coeur ou Pix+ sans encombre.

Commit:
refactor(api): filter on status in SQL query instead of using JS filter

yaf committed:
:sparkles: api(certif) add CalibratedChallenge

Co-authored-by: Alexandre Coin <alexandre@mbpdealexandre-1.home>

Commit:
feat(api): use calibrated challenge in get-next-challenge

Commit:
feat(api): add getManyCalibratedChallenges method

pix-bot-github commented:
Choisir les applications à déployer :

- [x] [API](https://api-pr14200.review.pix.fr/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr14200/environment) <!-- pix-api-review -->
- [x] [App (.fr)](https://app-pr14200.review.pix.fr) – [App (.org)](https://app-pr14200.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-app-review-pr14200/environment) <!-- pix-app-review -->
- [ ] [Orga (.fr)](https://orga-pr14200.review.pix.fr) – [Orga (.org)](https://orga-pr14200.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-orga-review-pr14200/environment) <!-- pix-orga-review -->
- [x] [Certif (.fr)](https://certif-pr14200.review.pix.fr) – [Certif (.org)](https://certif-pr14200.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-certif-review-pr14200/environment) <!-- pix-certif-review -->
- [ ] [Junior](https://junior-pr14200.review.pix.fr) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-junior-review-pr14200/environment) <!-- pix-junior-review -->
- [ ] [Admin](https://admin-pr14200.review.pix.fr) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-admin-review-pr14200/environment) <!-- pix-admin-review -->
- [ ] [API MaDDo](https://pix-api-maddo-review-pr14200.osc-fr1.scalingo.io/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-maddo-review-pr14200/environment) <!-- pix-api-maddo-review -->
- [ ] [Audit Logger](https://pix-audit-logger-review-pr14200.osc-fr1.scalingo.io/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr14200/environment) <!-- pix-audit-logger-review -->

> [!IMPORTANT]
> N'oubliez pas de déployer l'API pour pouvoir accéder aux fronts et/ou à l’API MaDDo.

Commit:
refactor(api): move calibrated challenge retrieval to dedicated repository

Commit:
refactor(api): rename calibrated challenge repository methods

alexandrecoin commented on api/src/certification/evaluation/domain/models/CalibratedChallenge.js:
```diff
@@ -0,0 +1,29 @@
+    this.colorBlindnessCompatibility = colorBlindnessCompatibility;
+  }
+
+  get isAccessible() {
```
Possible de retirer la méthode de Challenge.js ?

alexandrecoin commented on api/src/shared/infrastructure/repositories/challenge-repository.js:
```diff
@@ -116,6 +118,27 @@ export async function findValidatedBySkills(skills, locale) {
   return challengesDtosWithSkills.map(([challengeDto, skill]) => toDomain({ challengeDto, skill }));
 }
 
+export async function findActiveFlashCompatibleCalibratedChallenges({
```
Possible de retirer `findActiveFlashCompatible` ?

alexandrecoin commented on api/tests/tooling/domain-builder/factory/build-calibrated-challenge.js:
```diff
@@ -0,0 +1,17 @@
+  id = 'recCHAL1',
+  discriminant = 1,
+  difficulty = 0,
+  blindnessCompatibility = 'OK',
```
Utiliser l'énum `Accessibility`

alexandrecoin commented on api/src/certification/evaluation/domain/models/CalibratedChallenge.js:
```diff
@@ -11,13 +11,25 @@ export class CalibratedChallenge {
+    blindnessCompatibility,
+    colorBlindnessCompatibility,
+    competenceId,
+    skill,
```
Utiliser le skill.competenceId plutôt que le porter dans le challenge

alexandrecoin commented on api/src/certification/evaluation/domain/usecases/get-next-challenge.js:
```diff
@@ -72,23 +71,23 @@ const getNextChallenge = async function ({
 
-  const challenges = [...new Set([...alreadyAnsweredChallenges, ...activeFlashCompatibleChallenges])];
+  const challenges = [
+    ...new Set([...alreadyAnsweredCalibratedChallenges, ...activeFlashCompatibleCalibratedChallenges]),
```
answeredCalibratedChallenges

alexandrecoin commented on api/src/certification/evaluation/infrastructure/repositories/calibrated-challenge-repository.js:
```diff
@@ -0,0 +1,133 @@
+  return challenge1.id < challenge2.id ? -1 : 1;
+}
+
+function toCalibratedChallengeDomain({ challengeDto, skill }) {
```
renommer en _toDomain

alexandrecoin commented on api/tests/tooling/domain-builder/factory/index.js:
```diff
@@ -271,6 +272,7 @@ const certification = {
   },
   evaluation: {
     buildCalibratedChallenge,
+    buildCalibratedChallengeSkill,
```
Voir dossier scopé pour migrer

Commit:
refactor(api): move Accessibility enum from Challenge to CalibratedChallenge

Commit:
refactor(api): delete findFlashActiveCompatible method

Commit:
refactor(api): rename _toDomain method

Commit:
refactor(api): rename variables in get-next-challenge.js

Commit:
refactor(api): delete competenceId from CalibratedChallenge

Commit:
refactor(api): move CalibratedChallenge and CalibratedChallengeSkill builders to certif folder

Jeyffrey reviewed (approved):
<img width="289" height="360" alt="image" src="https://github.com/user-attachments/assets/0feba4c9-9694-44bf-bf4b-b69e8ef9e7ee" />

yaf reviewed (approved):
👌

Steph0 commented on api/src/certification/evaluation/domain/models/CalibratedChallenge.js:
```diff
@@ -0,0 +1,35 @@
+ */
+export class CalibratedChallenge {
+  /**
+   * Constructeur d'épreuve calibrée
```
```suggestion
```

Steph0 commented on api/src/certification/evaluation/domain/models/CalibratedChallenge.js:
```diff
@@ -0,0 +1,35 @@
+   * @param difficulty
+   * @param blindnessCompatibility
+   * @param colorBlindnessCompatibility
+   * @param skill
```
La JSDoc sert un peu à rien si on met pas les types sur les params 😅

Steph0 commented on api/src/certification/evaluation/domain/models/CalibratedChallenge.js:
```diff
@@ -0,0 +1,35 @@
+   * @param colorBlindnessCompatibility
+   * @param skill
+   */
+  constructor({ id, discriminant, difficulty, blindnessCompatibility, colorBlindnessCompatibility, skill } = {}) {
```
Attention, le `{}` indique que basiquement tous les paramètres sont optionnels, je crois comprendre en lisant la PR que ce n'est pas le cas et que beaucoup (si pas tous les champs) sont mandatory.

Pour éviter un objet anémique il faut que le model soit un moyen de comprendre les aspects métier de ce qu'il représente. Les deux modèles créés dans cette PR sont un peu léger sur ce point, en particulier au vu de leur rôle central pour le fonctionnel de Certif 😅 

```suggestion
  constructor({ id, discriminant, difficulty, blindnessCompatibility, colorBlindnessCompatibility, skill }) {
```

Steph0 commented on api/src/certification/evaluation/domain/models/CalibratedChallenge.js:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Traduction: Épreuve calibrée
+ */
+export class CalibratedChallenge {
```
```suggestion
export class CalibratedChallenge {
```

Steph0 commented on api/src/certification/evaluation/infrastructure/repositories/calibrated-challenge-repository.js:
```diff
@@ -0,0 +1,133 @@
+  );
+}
+
+function decorateWithCertificationCalibration({ validChallengeDtos, certificationChallenges }) {
```
Il ne faut pas récupérer cela de l'ancien code. Cette fonction ne sert plus à rien car on fait
* SELECT sql discriminant, difficulty
* alpha = discriminant => `alpha: discriminant` ci-dessous
* pour refaire discriminant = alpha => `discriminant: challengeDto.alpha` dans le toDomain

Steph0 replied:
merci !

Steph0 commented on api/src/certification/evaluation/infrastructure/repositories/calibrated-challenge-repository.js:
```diff
@@ -0,0 +1,133 @@
+
+  const certificationChallengeIds = certificationChallenges.map(({ challengeId }) => challengeId);
+
+  const findCallback = async (knex) => {
+    return knex
+      .whereIn('id', certificationChallengeIds)
```
Proposition d'ajouter le select afin de ne pas récupérer tout un ensemble de choses qui ne servent pas au nouveau `CalibratedChallenge` notamment et en particulier la alpha et delta.

Steph0 commented on api/src/certification/evaluation/infrastructure/repositories/calibrated-challenge-repository.js:
```diff
@@ -0,0 +1,133 @@
+  return challengesDtosWithSkills.map(([challengeDto, skill]) => _toDomain({ challengeDto, skill }));
+}
+
+async function _findChallengesForCertification({ versionId, locale, cacheKey, dependencies }) {
+  const certificationChallenges = await knex
+    .from('certification-frameworks-challenges')
```
Je propose de préciser le select (par exemple on a pas besoin de rapatrier pour 3000 challenges le `createdAt`) ?

Steph0 reviewed (approved):
Il est tellement moins fat le modele quand on prend que ce que nous avons besoin :D !

PS : vu l'heure j'ai fais que une relecture tech

Commit:
refactor(api): update JSDoc in models

Commit:
refactor(api): update calibrated challenge repository SQL select methods

Commit:
feat(api) use answered challenge for getMany method

Commit:
feat(api) deduplicate challenges

As AnsweredChallenges and CalibratedChallenges objects differ
the `new Set()` method used to deduplicate the array no longer works
A new method has to be created so that in case of duplicate object
(= when challengeIds are equal) only the object with the most properties
remain

alexandrecoin commented on api/src/certification/evaluation/infrastructure/repositories/answered-challenge-repository.js:
```diff
@@ -0,0 +1,55 @@
+  return challengesDtosWithSkills.map(([challengeDto, skill]) => _toDomain({ challengeDto, skill }));
+}
+
+async function loadChallengeDtosSkills(challengeDtos) {
```
mettre des tirets

alexandrecoin commented on api/tests/certification/evaluation/unit/domain/usecases/get-next-challenge_test.js:
```diff
@@ -979,4 +984,27 @@ describe('Unit | Domain | Use Cases | get-next-challenge', function () {
     });
   });
+
+  describe('#deduplicateByKeepingMostCompleteChallenges', function () {
```
deduplicateChallenges

alexandrecoin replied:
Méthode supprimée pour être remplacée par `uniqBy` de lodash.

Steph0 commented on api/src/certification/evaluation/domain/usecases/get-next-challenge.js:
```diff
@@ -158,4 +157,18 @@ const _getValidatedLiveAlertChallengeIds = async ({ assessmentId, certificationC
+    challenges.reduce((acc, challenge) => {
+      const existing = acc[challenge.id];
+
+      if (!existing || Object.keys(challenge).length > Object.keys(existing).length) {
```
`Object.keys(challenge).length > Object.keys(existing).length` j'ai du mal a comprendre ce cas, je veux bien que tu me l'enseigne.

Je propose de mettre cela dans une fonction nommee pour justement nommer ce cas.

alexandrecoin replied:
Retiré.

Steph0 commented on api/src/certification/evaluation/infrastructure/repositories/answered-challenge-repository.js:
```diff
@@ -0,0 +1,55 @@
+  const localeChallengeDtos = locale
+    ? challengeDtos.filter((challengeDto) => challengeDto.locales.includes(locale))
+    : challengeDtos;
+  localeChallengeDtos.sort(_byId);
```
Je me dis que le sorting peut etre integre a l'appel LCMS comme cette fonction contient deja 3 boucles et je me dis que PG peut faire le taf de sorting (voir meme de callback)

Un peu comme `findActiveFlashCompatible` finalement

alexandrecoin replied:
Vu ensemble, sera fait dans une autre PR.

Steph0 commented:
En  vrai je propose de faire tout ce qui est "opti" dans un ticket a part, le travail ici est deja consequent :)

Agnes-V reviewed (approved):
Test func fait avec @EmmanuelleBonnemay et @alexandrecoin pour Pix transverse et Pix+ Droit ✅

pix-service commented:
:tada: This PR is included in version 5.266.0 :tada:

The release is available on [GitHub release](https://github.com/1024pix/pix/releases/tag/v5.266.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
